File: 3_Sum.cpp
class Solution {
public:    
    /**
     * @param numbers : Give an array numbers of n integer
     * @return : Find all unique triplets in the array which gives the sum of zero.
     */
    vector<vector<int> > threeSum(vector<int> &nums) {
        // write your code here
        // sort(nums.begin(), nums.end());
        // vector<vector<int>> ret;
        // int length = nums.size();
        // int start, end;
        // for (int i = 0; i < length; ++i){
        //     if (i >  0 && nums[i] == nums[i - 1]) continue;
        //     start = i + 1;
        //     end = length - 1;
        //     while (start < end){
        //         if (nums[start] + nums[end] + nums[i] == 0){
        //             ret.emplace_back(vector<int>{nums[i], nums[start], nums[end]});
        //             ++start;
        //             --end;
        //             while (start < end && nums[start] == nums[start - 1]) ++start;
        //             while (end > start && nums[end] == nums[end + 1]) --end;
                    
        //         }else if (nums[start] + nums[end] + nums[i] > 0){
        //             --end;
        //         }else{
        //             ++start;
        //         }
        //     }
        // }
        // return ret;
        // better codes of the same solution
        sort(nums.begin(), nums.end());
        vector<vector<int>> ret;
        int length = nums.size();
        int start, end;
        for (int i = 0; i < length; ++i){
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            start = i + 1;
            end = length - 1;
            while (start < end){
                if (start > i + 1 && nums[start] == nums[start - 1]){
                    ++start;
                }else if (end < length - 1 && nums[end] == nums[end + 1]){
                    --end;
                }else{
                    const auto sum = nums[i] + nums[end] + nums[start];
                    if (sum > 0){
                        --end;
                    }else if (sum < 0){
                        ++start;
                    }else{
                        ret.emplace_back(vector<int>{nums[i], nums[start], nums[end]});
                        ++start;
                        --end;
                    }
                }
                
            }
        }
        return ret;
    }
};

File: 3Sum.py
class Solution:
    """
    @param numbersbers : Give an array numbersbers of n integer
    @return : Find all unique triplets in the array which gives the sum of zero.
    """
    def threeSum(self, numbers):
        # write your code here
        if not numbers:
            return []
        numbers.sort()
        ans = []
        
        length = len(numbers)
        for i in range(length):
            if i > 0 and numbers[i] == numbers[i - 1]:
                continue
            left, right = i + 1, length - 1
            while left < right:
                total = numbers[i] + numbers[left] + numbers[right]
                if total > 0:
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    ans.append([numbers[i], numbers[left], numbers[right]])
                    left += 1
                    while left < right and numbers[left] == numbers[left - 1]:
                        left += 1
                    right -= 1
                    while left < right and numbers[right] == numbers[right + 1]:
                        right -= 1
        return ans